### ISSUE TITLE
MPD manifest error causes playback delay

### ISSUE THREAD

**Issue by dagwieers** [peak3d/inputstream.adaptive#433]
We discovered that an incorrect MPEG-DASH manifest, including a subtitle that gives HTTP 415 (Unsupported Media Type) cause IA to retry downloading 10 times causing playback delays from 10 to 20 seconds.

Maybe if a subtitle track causes HTTP 415, it should not be retried, but skipped.

See add-ons/plugin.video.vrt.nu#735 for the details.

**Comment by mediaminister** [comment]
> Maybe if a subtitle track causes HTTP 415, it should not be retried, but skipped.

If audio and video are okay, these streams should play immediately independent from the errors with subtitles.

**Comment by dagwieers** [comment]
@mediaminister Is HTTP 415 ever a recoverable error? To me it does not seem likely.

~**Update:** In fact, if I am not mistaken, in our case the problem seems to be that the subtitle track has content-type "application/mp4".~

**Comment by basilgello** [comment]
Same applies not only to subtitle but for HLS streams where server throws 404 Not Found on segment. Result is video looping the same 10 second segment again and again, making playback unwatchable.

@dagwieers @peak3d @mediaminister I guess **all HTTP 4xx return codes** must skip retry downloading the borked segment and move to the next segment. After, say, five consecutive segments can not be downloaded, the player should gracefully exit.

P.S: I sent the log file with a reproduced error to Markus by email two or three weeks ago, but seems moving into new apartment (if Markus's Github presence note is correct) did prevent him from reacting on the bug report.

**Comment by glennguy** [comment]
> I guess all HTTP 4xx return codes must skip retry downloading the borked segment and move to the next segment

The retry downloading is quite useful for streaming at the live edge, where sometimes server load issues will cause a segment (and even the playlist) to not yet be available/404. For this case skipping this segment and 'fast forwarding' to the next would be bad.

**Comment by dagwieers** [comment]
We got feedback from the stream provider (https://github.com/add-ons/plugin.video.vrt.nu/issues/735#issuecomment-625741342) with a technical explanation why this happens and why this is expected.

> The problem as I understand it is indeed what you already figured out above: Radio 1 (unlike our other radio channels) sometimes includes (closed, i.e. non-burned-in) subtitles. To support this case, the manifest includes a subtitle track. But theses subtitles are not used all the time: most of the time, there are no subtitles present. We can't leave these out of the manifest, because the player would need a full reload if we do need subtitles.
> 
> Since the subtitle track is not (always) present, the origin can't provide the requested data, and returns the 415 status that you observed. From the point of view of our origin partner, this is expected behaviour.
> 
> The player (TheoPlayer) we use on our official sites/apps ("touchpoints") handles this 415 as meaning "no subtitles present", and starts to play anyway. The players you tested (ffmpeg, jwplayer) behave differently. (Note that I'm trying to avoid pointing fingers at who's right or wrong)

I think we need to reconsider how to handle certain 4xx errors, potentially taking into account the content-type.

*The VOD provider is using [Unified Streaming Platform](https://docs.unified-streaming.com/introduction/index.html) ([Unified Origin -- LIVE](https://docs.unified-streaming.com/documentation/live/index.html)) and [THEOplayer](https://docs.portal.theoplayer.com/) as an embedded web-player.*

**Comment by mediaminister** [comment]
This causes the delay: https://github.com/peak3d/inputstream.adaptive/commit/cb3c12b97b07628d06d55cc4de07b782afec03a4

**Comment by dagwieers** [comment]
Reading the [DASH-IF Guidelines for Implementation: DASH-IF Interoperability Points V4.3](https://dashif.org/docs/DASH-IF-IOP-v4.3.pdf),

Page 86, chapter **4.3.6. Considerations on live edge**

> A DASH client should avoid being too aggressive in requesting segments exactly at the computed segment availability start time, especially if it is uncertain to be fully synchronized with the server. If the DASH client observes issues, such as 404 responses, it should back up slightly in the re9 quests.
>
> In addition, for a content authoring to avoid too aggressive requests and possible 404 responses, the content author may schedule the segment availability start time in the MPD with a small safety delay compared to the actual publish time. This also provides the content author a certain amount of flexibility in the publishing of Segments. However, note that such safety margins may lead to slightly increased end-to-end latencies, so it is a balance to be taken into account.

Page 105, chapter **4.8. Robust Operation**:

> - If the DASH access client receives an HTTP client error (i.e. messages with 4xx error code) for the request of an Initialization Segment, the Period containing the Initialization Segment may not be available anymore or may not be available yet. 
> - Similarly, if the DASH access client receives an HTTP client error (i.e. messages with 4xx error code) for the request of a Media Segment, the requested Media Segment may not be available anymore or may not be available yet. In both these case the client should check if the precision of the time synchronization to a globally accurate time standard or to the time offered in the MPD is sufficiently accurate. If the clock is believed accurate, or the error re-occurs after any correction, the client should check for an update of the MPD. . If multiple BaseURL elements are available, the client may also check for alternative instances of the same content that are hosted on a different server

Page 106, chapter **4.8.5. Segment Unavailability**

> To address signaling of segment unavailability between the client and server and to indicate the reason for this, it is recommended to use regular 404s. In addition, unless a UTC Timing has been defined prior in the MPD, the Date-Header specifying the time of the server should be used. In this 
 case, the DASH client, when receiving a 404, knows that if its time is matching the Date Header, then the loss is due to a segment loss.

Annex B, Page 5, chapter **B.4.5 Segment Unavailability**

> When a server cannot serve a requested segment it gives an HTTP 404 response. If the segment URL is calculated according to the information given in the MPD, the client can often interpret the 404 response as a possible synchronization issue, i.e. its time is not synchronized to the time of39 fered in the MPD.
>
> In the MBMS case, a 404 response is also likely to be caused by non-reparable transport errors. This is even more likely if it has been possible to fetch segments according to the MPD information  earlier. Although the client M/W, which is normally located in the same device as the DASH player, knows what segments have been delivered via broadcast and which ones are missing in a sequence, it cannot indicate this to the DASH client using standard HTTP responses to requests for media segments.

It references the HTTP/1.1 specification ([RFC 7231](https://tools.ietf.org/html/rfc7231)) for the appropriate action on errors, whereas RFC 7231 does not specify actions per HTTP 4xx error. It does mention:

> Unless noted otherwise, a recipient MAY attempt to recover a usable protocol element from an invalid construct.  HTTP does not define specific error handling mechanisms except when they have a direct impact on security, since different applications of the protocol require different error handling strategies.  For example, a Web browser might wish to transparently recover from a response where the Location header field doesn't parse according to the ABNF, whereas a systems control client might consider any form of error recovery to be dangerous.

**Comment by mediaminister** [comment]
I made a draft PR which disables retrying for http error 415: https://github.com/peak3d/inputstream.adaptive/pull/441 (which fixes https://github.com/add-ons/plugin.video.vrt.nu/issues/735)

But before disabling retrying for other http errors, I hope @peak3d can explain why the commit that introduces the retrying mechanism for segment downloads https://github.com/peak3d/inputstream.adaptive/commit/cb3c12b97b07628d06d55cc4de07b782afec03a4 is needed.

**Comment by glennguy** [comment]
@mediaminister I know of at least 1 add-on that relies on the robustness of having 404 retries for livestreaming, there are possibly others. In my case I've seen that segments close to the live edge (and the playlist refreshing itself btw) will intermittently 404 during peak usage times. The official app/player doesn't have issue with it so they would likely have allowed for 404s to occur without the stream failing/exiting.

Maybe we could go for a more flexible solution where you could specify the amount of retries using a listitem property, and change the current 10 retries default to a more reasonable 3? Individual content providers can customize their player to suit their workflow and needs but I don't think with the vast array of content used in this add-on we can be so rigid.

**Comment by dagwieers** [comment]
@glennguy @matthuisman The proposed PR only skips retrying for subtitle tracks that give a HTTP 415 error.

Making retries configurable for HTTP errors seems the wrong thing to do. Some things are recoverable, others are not. It makes no sense to retry stuff that is not ever recoverable.

For that we obviously need to test and document the behaviour we see in the field, not base our judgement on past encounters. or hearsay. We should establish what is correct behaviour and compare this to other players (and reference implementations) as well.

Now Kodi/AI acts poorly, both for starting streams (it takes 4 to 5 seconds for every stream, which makes channel hopping impossible) as well as handling unexpected behaviour (like stalling for 10 to 20 seconds on a subtitle track that will never be there (415) before starting playback).

**Comment by matthuisman** [comment]
why am I tagged?

Pretty sure @glennguy  was replying to the other comment (not this issue / PR specifically) about a possible reason retry logic was added. But I guess only Mr peak can confirm.

Remember that the issue here is the "bad mpd" listing subtitles that don't exist.
Seems a lot of IA code is related to services not doing things correctly.

We have a workaround for slow services (retry logic)
Now we want to remove that logic to fix services that have bad mpds.

Which is correct? Well, neither.
So lets try to find a happy medium where as many work as possible without hacking too much.

We could all agree 415 should not retry?
But if we have edge case for 415, what about 500, 403, 405 etc

I think maybe better is just not to retry on subtitles fullstop?
If response not ok and is subtitles - don't retry.
Then don't need to specifically look for 415.
Subtitles not working isn't breaking playback so aren't "crucial"

So,
`if (type_ == AdaptiveTree::SUBTITLE && returnCode == 415)`
becomes
`if (type_ == AdaptiveTree::SUBTITLE && returnCode != 200)`

Oh, and I agree with Glenn on at least reducing the retries to 3 instead of 10.

**Comment by dagwieers** [comment]
> why am I tagged?

You and @glennguy seem to be pretty ~one-minded~ like-minded about things. The response underlines this once more :-)

> We have a workaround for slow services (retry logic)
> Now we want to remove that logic to fix services that have bad mpds.

Nobody is proposing to remove it altogether. The PR is pretty strict, and the proposal was to understand the rationale of the current implementation.

> Remember that the issue here is the "bad mpd" listing subtitles that don't exist.
> Seems a lot of IA code is related to services not doing things correctly.

If you read the follow-up, it is not necessarily considered a bad mpd, and the streaming back-end (and player) supports this purposely. So again, I would prefer we list known cases and document the rationale for future behaviour so we do not get into a ping-pong where person A fixes his issue, and 6 months later person B fixes her issue, breaking someone else's use-case.

Preferably we list the streams/back-ends, test them against other players and document behaviour and outcome.

**Comment by matthuisman** [comment]
whats with the personal attacks?
Have I done something to you??

You tagged me for some reason - I'm not the developer of IA... I do PRS just like you do...
I hadn't commented on this PR at all until you tagged me.
So was simply asking why I was tagged / brought into this conversation.

i think there is some language barrier.
It seems you think I'm arguing.
I'm just explaining how I see things and having a discussion.

I'm not even against this PR.
We are having a general conversation about retry logic at this point

**Comment by dagwieers** [comment]
From [DASH-IF Live Media Ingest Protocol](https://dashif-documents.azurewebsites.net/Ingest/master/DASH-IF-Ingest.html) related to 4xx errors:

### 4. General Ingest Protocol Behavior
> 8. In case of an authentication error confirmed by an HTTP 403 response, the ingest source SHALL retry establishing the Connection within a fixed time period with updated authentication credentials, when that also results in error the ingest source can retry N times, after this the ingest source SHOULD stop and log an error. The number of retries N MAY be configurable in the Ingest Source.

> 14. The Ingest source SHOULD resend Objects for which a connection was terminated early, or when an error response was received such as HTTP 400 or 403 if the connection was down for less than 3 average segments durations. For connections that were down longer, ingest source can resume sending Objects at the live edge of the live media presentation instead.

> 17. In case the Receiving entity cannot process the POST request due to authentication or permission problems, or incorrect path, then it SHALL return a permission denied HTTP 403 with error reason
>
> 18. In case the Receiving entity can process the fragment in the POST request body but finds the media type cannot be supported it MAY return an HTTP 415 unsupported media type, otherwise 400 bad request MUST be returned.
>
> 19. In case an unknown error happened during the processing of the HTTP POST request a HTTP 400 Bad request SHALL be returned by the Receiving entity
>
> 20. In case the receiving entity cannot process a fragment posted due to missing or incorrect init fragment, an HTTP 412 unfulfilled condition MAY be returned, otherwise, in case this is not supported by the system, a HTTP 400 bad request response MUST be returned.

> 22. In case the receiving entity or publishing point receiving the HTTP POST body is not setup or available an HTTP 404 not found SHOULD be returned to the ingest source.

### 6.1.6. Additional Failure behaviors
> 2. Upon receipt of an HTTP 403 or 400 error, the ingest source MAY be configured to NOT retry sending the fragments, hence N will be 0, as described in general

**Note:** This is not the communication between player and CDN, but rather the Live encoder and i.e. the CDN.

**Comment by dagwieers** [comment]
What personal attack? I am not arguing, neither do I think you are.

**Comment by matthuisman** [comment]
> You and @glennguy seem to be pretty one-minded about things. The response underlines this once more :-)

This is saying I'm one-minded and my response was also.
What else have I been one-minded about?

**Comment by dagwieers** [comment]
Ok, I meant you both are like-minded. (Mind you English is not my native language)

**Comment by dagwieers** [comment]
The reason I added you was you liked @glennguy's comment minutes after he made it, so I figured you both had discussed this, like with many other conversations. I am not sure what I did wrong with tagging you. I also don't see why that would be offensive.

**Comment by matthuisman** [comment]
Ok ok, that makes more sense.

Language barriers can seem quite bad.
Like how you quote this

```
We have a workaround for slow services (retry logic)
Now we want to remove that logic to fix services that have bad mpds.
```

To you, that seems like I'm proposing that.
But I'm not - I'm just "thinking out loud" about how silly thing can become.
A big round-about of fixing / unfixing to suit each service.

And I quote "bad mpd" as I'm not meaning bad. But just a way to describe a mpd that contains a link to something that isn't actually present.

if all services were 100% "good" we wouldn't need any of these workarounds.

**Comment by basilgello** [comment]
Would love to hear from @peak3d & @phunkyfish here as well :)

**Comment by dagwieers** [comment]
> if all services were 100% "good" we wouldn't need any of these workarounds.

The problem is that the specification is not 100% clear about behaviour, but we do know that existing infrastructure works with existing video players, where Kodi/IA does not.

This is why we need to document known/existing behaviour of infrastructure (preferably with as much details as possible) so that in the future we can easily retest corner cases and how Kodi/IA behaves. That is the only way we can improve Kodi/IA without introducing regressions.

(And we need to test preferably each release to give everybody a chance to do real life testing before it hits the mirrors and impacts users...)

**Comment by matthuisman** [comment]
sounds good :)

I have actually started a repo today adding different files that I can test IA with.
As I found Matrix has a regression of IA HLS with AES128 playback if it uses multiple keys.

https://github.com/matthuisman/ia_tests

tests.json could be pulled by a Testing add-on to allow for quick testing inside Kodi
It just lists a name for the test (list item label), then the properties to set and the url.

Anyway, back to topic on this Issue...

What does the spec say?
Do you think subtitles should be retried if the code isn't 415 eg. 404?

**Comment by dagwieers** [comment]
We are documenting tools for troubleshooting and specifications here: https://github.com/add-ons/best-practices/wiki/Streaming-issues (it also includes known streams for testing specific behaviour/support)

I was also considering writing an add-on with the sole purpose of (automatically) testing behaviour in Kodi. Where we can test seeking, pausing, etc. And where test-files can be selected from a list, or fed directly into the add-on.

This would greatly help in having people troubleshoot their issues, and report back.

**Comment by matthuisman** [comment]
awesome

I usually refer to Shaka-player codebase to see how they do things as that is used a huge amount Which means lots of stackoverflow questions / answers. Documentation etc. 
https://github.com/google/shaka-player

Example of how they do their retry logic
https://github.com/google/shaka-player/blob/master/docs/tutorials/network-and-buffering-config.md



OK, I tried your file with all those other players - and none seem to work correctly.
Some retry a few time and give up completely.
Some keep re-downloading the mpd after they fail and start cancelling requests etc.
Some just ignore subs so work ok.

IA seems to be doing ok - at least it works after initial slow start.

If we want to keep the same behaviour as other players - then this stream is still not going to play correctly.

Need to decide if we want to keep to the spec and copy what other players do, or have custom behaviour to improve the user experience / support more of our addons?

If we remove retry logic on 415 subs - that's custom behaviour to IA only.
It solves the issue but is it right?

Can't have 100% to spec and have 100% compatibility.

I think a good start is reduce the retries from 10 to 2 (or 3)
Maybe also remove / reduce the sleep between retries.

My personal opinion is - screw the red-tape.
Lets do what is best for the end-user.

I think subs are an enhancement not a requirement.
I don't think we should slow down playback waiting for them.
give them the initial request - and if they are not OK - then just return.
ie. Don't retry on subs.

**Comment by matthuisman** [comment]
So, what do other players do in regards to 415 subtitle behaviour?

**Comment by glennguy** [comment]
Hey @dagwieers sorry for changing the topic from the core of this issue - I was just responding to mediaminister's comment about whether to remove the retry mechanism as suggested by @basilgello 

I can imagine a very simplified process for content providers in choosing a CDN for their content, a player for each platform, then configuring both to work together well. Some of the players out there are open source and the content provider developers have scope to customise things further to their needs.

In our situation however (well particularly when DRM is involved) within the Kodi ecosystem we only have a choice of one player - inputstream.adaptive . Even though it's open source, because we don't/can't have a binary addons repository that works across all platforms (looking at you Android in particular), so running our own forks of IA is not practical either.

Also to keep in mind that not only should we look to the DASH IF recommendations but also keep in mind that these could differ to smoothstreaming and HLS. For example in the current [HLS draft](https://tools.ietf.org/html/draft-pantos-hls-rfc8216bis-06) it says:
> The absence of media data (due to, for example, the temporary
   unavailability of an encoder) SHOULD be signaled....

the key word being 'SHOULD' (their emphasis). There are not recommendations as to what the client should/must do in this case though. The code that handles downloading is generalised for all 3 supported playlist types so any change will affect all 3.

@dagwieers I agree that we shoudn't blanket retry all 4xx errors - 404 we should and 403 we have the trigger to renew token/auth, but yes unrecoverable errors like 400 should not be retried, and probably many more 4xx errors where this is the case.

@mediaminister 's PR looks like a great way to solve the issue in the OP, and provide for further extensibility with other media/error combinations. My opinion moving forward is that since the current code is a very lenient approach to error handling we should be careful in 'winding' it back to something more rigid i.e 'if it isn't broke...'

**Comment by basilgello** [comment]
Thanks for clarificatiin and mention @glennguy! On my side I try to investigate is it useful at all to move to next segment on 404. I agree totally removing one edge case code to support another edge case is not wise. However, I'd like to make retry count configurable. Like, for user watching an IPTV provider with HLS streams mostly, the sane setting will be 3 retries or even one. If my memory serves me well, after retry fails the next segment is queried anyway so in this case reducing retry count would be sufficient. If not, the setting whether to try load next segment on 404 or bail out can be configurable too.

I can prepare a PR making these things configurable. And it is very good that we try to define a single table of actions needed to handle different non-2xx codes.

**Comment by dagwieers** [comment]
@matthuisman Yes, other players don't handle those cases very well either. But some do, and the ones supported by the VOD provider handle this without issue. IA is competing with the stuff that VOD providers offer in their browser experience. And VOD providers do not support Kodi/IA, so we will have to handle these cases as well (or find compromises to support as many cases where needed).

@glennguy We need to document ASAP all the cases that do require retries, and start listing everything we want to test in this regard. Because even change the retries from 10 to 3 is not something we know for sure does not break anything, if we don't know why this was added in the first place. (Why always retry? Why 10 times? We have no clue...)

**Comment by dagwieers** [comment]
BTW when we support different quality in streams a 404 error could retry the same segment on a lower quality stream, on success return to the next segment in the original stream. This could offer recoverability in cases where it would be unrecoverable. This is how some other players do this, not just for 404 errors, but also on connection issues or corruption. Rather than retry, always try the alternative(s) first before retrying the original.

**Comment by matthuisman** [comment]
The GSOC for adaptive streaming in IA should hopefully take care of that.
Wouldn't help your case though as there is only a single represention for the subs

We know why retry logic was added.
Because nearly every other player does it.

And I'm a little confused.

You seem to have changed your thoughts.

You originally wanted IA to follow the spec, follow what other big players do. And now you seem to want custom logic to suit a single vod player? Their developer even said it was a workaround. 

You seemed angry IA added something that is in every other player that made your one stream delay by 10 seconds when it doesn't even play in majority of other cases. You should be congratulating it as its doing a better job than most.

You can't expect IA to test every single stream before implementing a feature that is actually widely used. 

Again, I'm happy for custom logic if we can agree on what has the most wide uses. I agree not to always follow spec or copy other major players. I agree that sub's shouldn't retry. They have a chance to fix themselves in future requests. Let's just play the video.

I just don't agree with how you've handled this discussion.

You came in like your fixing a bug / regression in IA.
 But really its a feature request to support your stream, so try be nicer.

**Comment by dagwieers** [comment]
> We know why retry logic was added. Because nearly every other player does it.

That is clearly not the case, and not 10 retries causing delays up to 20 seconds before starting.

> You seemed angry

I am not.

> You came in like its an issue / bug in IA.

I did not. Read the original report.

> But really its a feature request to support your stream, so try be nicer.

Not sure what you are insinuating now. This is about improving IA, I don't care if handling unexpected cases better is a bug-report or a feature request. That should not matter. The goal is to have a better working product, and I assume you are here for the same reasons.

**Comment by matthuisman** [comment]
I am. But we can't have a better product if we follow the major players or spec.

You said to look at what other players do.
If we follow them, your stream won't play at all. 

I think of IA as kodis web player. We can add custom logic to support kodi streams like how your provider does for theirs.

But the issue is that kodi has many streams from many providers all doing their own stuff as they only need to support their own players.

And we only know what our addons need. 
Could be a bunch of people using .strm files to play their providers mpds.

Again, I'm for skipping subtitle retries on error. But not specific errors. I'm also for reducing the amount of retries to be more inline with other players. And for making that a property that can be adjusted if needed.

**Comment by dagwieers** [comment]
> You said to look at what other players do. If we follow them, your stream won't play at all.

Depends on which players you select. If you select the ones that cannot handle the same unexpected cases, yes. Should we follow them? I would hope not. The specification allows for stopping playback on unexpected issues, but that would get us a lousy and brittle player.

I don't think we are in disagreement. I simply want all edge cases to be documented, and retries to be performed where the actions are recoverable. Both are clearly not the case today. I don't know why retries is set to 10 and why they are performed for all error cases.

**Comment by dagwieers** [comment]
> But the issue is that kodi has many streams from many providers all doing their own stuff as they only need to support their own players.

I don't think this is the case, these are often the same commercial backends and players on the market. They may have issues, but we are not even talking about those, we are talking about undocumented behaviour and handling it sanely. That is what the PR is about IMO.

**Comment by matthuisman** [comment]
I tried the players you link to in your readme for testing streams.

415 error is just random. The provider probably picked it so they could tell difference vs a 404 in their player. Its not a known error to skip retry.

Nothing about a 415 error tells me not to retry. If we didn't have this vod service, would we know not to retry on 415?

If anything, 404 shouldn't retry as that means it doesn't exist. However we know with live edges (and other players), that we need to retry on this error. 

Its the internet after all. Errors happen.

So I think we should just treat subtitles different. As if we had a 415 on video we still want to retry?

I like the "if sub's fails, just skip them for now" vs "if we get this specific error code from one provider, skip"

Or could have a subs_retry property as well. 

As I think subs should be treated different as they arnt 'essential' for playback. Better to have a video without sub's than no video or delayed video.

And sub's can always fix themselves on next go

**Comment by matthuisman** [comment]
> > But the issue is that kodi has many streams from many providers all doing their own stuff as they only need to support their own players.
> 
we are talking about undocumented behaviour and handling it sanely. That is what the PR is about IMO.

Its undocumented because only your provider does it. 

> > But the issue is that kodi has many streams from many providers all doing their own stuff as they only need to support their own players.
> 
> I don't think this is the case, these are often the same commercial backends and players on the market. They may have issues, but we are not even talking about those, we are talking about undocumented behaviour and handling it sanely. That is what the PR is about IMO.

And yet your one providers streams don't work on any of those major platforms.  

Its undocumented as the one provider who does it hasn't documented it.

**Comment by dagwieers** [comment]
- **[HTTP 404 Not Found](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404)** means the file is not available, but the specification clearly stipulates that it may not be available yet and suggests *the client should check if the precision of the time synchronization to a globally accurate time standard*. So yes, this may be recoverable for the exact case @glennguy coined.
- **[HTTP 415 Unsupported Media Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/415)** is an unrecoverable error, nothing suggests doing the exact same request again would make it suddenly *a supported media type*. In that regard it looks like **[HTTP 400 Bad Request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400)**, not sure why they needed to differentiate from a 400 in this case. But then again, HTTP responses are lacking mostly in the specification.

**Comment by matthuisman** [comment]
The developer said the files arnt there. So why use 415? What happens when they change it to 404 instead? 

Can you link one other player that handles it correctly?

**Comment by dagwieers** [comment]
@matthuisman I am getting tired of this discussion. We seem to be going in circles.

> Its undocumented because only your provider does it.

It is not *my* provider, it is hard to proof that only this provider does it, and it is really not documented how the behaviour should be, as such undocumented. Apparently this provider considers this the intended behaviour.

> And yet your one providers streams don't work on any of those major platforms.

~The ones I listed are the ones that have a web interface, I wouldn't say these are the major platforms used in the field by large corporations.~ **Updated**: I added THEOplayer, and apparently your statement is incorrect, this *faulty* streams works on 5 out of 7 listed players. But indeed, the specification allows to stop handling a stream on unexpected errors (even on HTTP 400 and HTTP 404 for that matter). 

(*Provider* is the wrong entity here, because this broadcaster makes use of well-known commercial software for encoding/transcoding and web-player, it's not in-house developed stuff)

> Can you link one other player that handles it correctly?

The player the provider is using is [THEOplayer](https://demo.theoplayer.com/test-your-stream-with-statistics), which I would consider one of the major players.

[Shaka Player](https://shaka-player-demo.appspot.com/demo/#panel=CUSTOM%20CONTENT) plays it without issues as well.

The [DASH references player](https://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html) also plays it without issues, but reports the subtitle track is missing. As well as [Akamai's player](https://players.akamai.com/players/dashjs)

And the [Bitmovin player](https://bitmovin.com/demos/stream-test) also plays without issues.

So from [my list](https://github.com/add-ons/best-practices/wiki/Streaming-issues), only **FlowPlayer** and **JW Player** have issues playing this stream with the HTTP 415 error on subtitles. Are these the **major platforms** you were talking about, Matt?

Mind you, IA handles it reasonably well since it eventually plays, but that behaviour could be improved as HTTP 415 clearly is intended to be unrecoverable.

**Comment by dagwieers** [comment]
BTW the reason why this problem is so well documented here is exactly because the provider and the vendor have been quite open on this. The technical reason to why this subtitle is not available and why this is intended. (This information was not available to me when I opened the ticket).

> The developer said the files arnt there. So why use 415? What happens when they change it to 404 instead?

(My guess) Because 404 is recoverable (according to MPEG-DASH specifications) whereas 415 is not (unspecified in MPD but clearly intended from the HTTP specs).

**Comment by matthuisman** [comment]
Did you even check the webtools when trying to play the streams?

Some don't even try to get subtitles (like theoplayer)
The others keep redownloading the mpd after every error on the subs and then sometimes cancel requests to video segments if half way through downloading the new mpd

So - no - they aren't working

The issue I have is with 415.
They are saying the format isn't supported - what format do they want?
If the file isn't there - they should return a status that tells you that.
They have done a "work-around" to stop them having to do the below.

If they don't have subs - remove the subs from mpd.
If they have subs that come and go - then return files with no subs in them.
This would allow subs to start half way through a segment as well.

So, can we stop relying on a 415 and work on a broader feature.

I think not retrying on subtitles on any error sounds good?
Subs shouldn't affect our playback or slow it down.
If they fail - then just skip them for now.

**Comment by dagwieers** [comment]
@matthuisman I am not going over this in circles, 5 out of 7 play the stream. THEOplayer supports subtitles just fine, but not in this case because they aren't available. Playback is smooth despite the HTTP 415.

**Comment by matthuisman** [comment]
No, your not looking at what the players are requesting!
**They are ignoring the subs so of course they work.**

Its not circles - your just not making any valid arguments.
They return 415 - incorrect format.
What format do they want?

Can you admit thats the wrong error to return? It makes 0 sense.

**Can you point my to a player that actually tries to request the subs but then keeps playing reliably?**

For anyone else who wants to join in, this is the mpd:
https://live-radio-cf.lwc.vrtcdn.be/groupa/live/d6fdec5d-5d67-4cc3-9f8c-e3e8993c8ee4/live.isml/.mpd

The subs will have **live-text_track_0_dut** in them

**Comment by matthuisman** [comment]
**Theo player (no requests for subs)**
![image](https://user-images.githubusercontent.com/6225961/81612483-0536b080-9431-11ea-9a25-1d77743aa1b3.png)
**Bitmovin (no requests for subs)**
![image](https://user-images.githubusercontent.com/6225961/81612641-416a1100-9431-11ea-95dc-129dc3f6172e.png)
**Shakka (no requests for subs)**
![image](https://user-images.githubusercontent.com/6225961/81612744-6d859200-9431-11ea-8c7d-2bd3312ce067.png)
**Akamia (does request subs)**
But then redownloads the mpd, cancels a request and retrys the same text track over and over
![image](https://user-images.githubusercontent.com/6225961/81612944-b9d0d200-9431-11ea-91cd-b04e80d4234f.png)

**Comment by matthuisman** [comment]
Bitmovin (no requests for subs)
![image](https://user-images.githubusercontent.com/6225961/81612641-416a1100-9431-11ea-95dc-129dc3f6172e.png)

**Comment by dagwieers** [comment]
@matthuisman Indeed THEOplayer doesn't appear to get the subtitles on the demo player, but it does support them on the VOD provider's website: https://www.vrt.be/vrtnu/livestream/

But Shaka player does fine, let me test the others. *(You need to enable subtitles in the player for some players)*

**Comment by matthuisman** [comment]
See my comment above with shakka - doesn't try subs either

Could you play the stream on their webplayer and paste in the requests it makes like I did above?
Maybe theoplayer test player is different to production?
Or maybe they have added in custom code on top of it (override / callback)

**Comment by dagwieers** [comment]
~So the only player that supports subtitles and works is the DASH reference player :smile: 
Four don't support subtitles in the demo version, and 2 fail to work when they do get the subtitles.~

~That means we only have 3 players to verify the behaviour of, and two are not very resilient to the issue we reported. Flowplayer and JW Player have code on GitHub so let me report this to them as well.~

So 4 out of 7 players play without issue when you enable subtitles (of which 2 report the issue without impacting playback). One stutters at the beginning but continues fine. 2 fail immediately.

**Comment by matthuisman** [comment]
Reference player still retrying the same subs (same as Akamia)
Its just happening via AJAX so doesn't affect playback.
IA is in the same thread
Actually, it retries 3 or 4 times then gives up (with the error)

![image](https://user-images.githubusercontent.com/6225961/81614891-d4587a80-9434-11ea-9c00-748c6e22497b.png)

**Comment by dagwieers** [comment]
Okay, so here is the THEOplayer from the VOD provider for this specific stream: https://radio1.be/kijk-live

You have to press the square box on the lower-right to enable subtitles (it starts with subtitles off).

It is in fact continuously trying to get new subtitle segments, which is what you would expect if the subtitles could become available at some point in the future. Which is the intended use-case for this.

But it doesn't affect playback.

**Comment by matthuisman** [comment]
because webplayers can use ajax (in different threads)

Is it trying the same subtitle file or the next one in line?
Why doesn't the provider just return a file with no subs in it?

**Comment by dagwieers** [comment]
@matthuisman Because it isn't necessary I guess, it works for them like this. (They were nice and told us they look into this). The intention here is to make IA work better despite unexpected behaviour. I don't think we can make Netflix, Disney+ or HBO GO to change their ways because we cannot handle their return codes.

So even if your content provider fixes this issue, I would still have reported it here because it is about resilience/reliability.

In that regard, I would be interested of how FlowPlayer and JW Player will tackle this issue. I guess they can retry asynchronously as well.

**Comment by matthuisman** [comment]
So, to confirm  -we are happy with the retry logic then (but maybe less retries)?
And now agree 415 is retried by everyone.

I guess the real fix would be IA using different threads for video / audio / subs.
That's a big job. 
But i think the GSOC adaptive may be using that to fetch future segments etc, so may sort it out

Maybe we can consider my idea of just skipping retries of subs alltogether?
It'll try the next sub next time it fetches mpd anyway.

**Comment by dagwieers** [comment]
I never was against retries, and I still think HTTP 415 is unrecoverable. We still don't know why they return 415 in this case, but at least THEOplayer doesn't mind at all. The reference player gives up, but doesn't impact playback.

So the best fix would be something that doesn't affect playback IMO. We won't have subtitles with the current behaviour (or if we drop retries) so the behaviour is suboptimal in any case.

**Comment by matthuisman** [comment]
The players do MIND.
They keep retrying the same file.
They are not skipping retries on 415

Yes, you get the playing result - but you can't just ignore what they are doing.
They are retrying - but it's in different threads.
IA is doing the same thing -  but only single thread.

So, we can either use different threads - or we can skip retries.
We can't have both

I vote for now to skip retries of subtitles altogether.
We don't want subs affecting our playback.
Just like JavaScript in a website, it should be an enhancement - not a requirement.
However, that was thrown out the window with web 2.0.

Or maybe subs only retry once. and video / audio retry 3 times.

Something like this I like more than specifically looking for 415.

But you say the subs may start working - so you do want it to retry eventually?
Does IA keep retrying after eventually starting playback (eg. after redownloading mpd)?

Have you asked the provider if they can just return a subs file with no subs in it?
Seems that would be the correct approach for subs that come and go.
That would make their MPD work in all the players.
It also saves all the extra requests on their server from their own player :)

It's listed in the MPD - as far as IA is concerned - it should be there - thats why it retries.
IA shouldn't need to think some MPD content is "optional".

**Comment by dagwieers** [comment]
Just as a matter of due diligence I reported the playback issues with JW Player and FlowPlayer upstream as well:
- JW Player: https://github.com/jwplayer/jwplayer/issues/3680
- FlowPlayer: https://github.com/flowplayer/flowplayer/issues/1461

So a small recap of all test players (from best to worse wrt. handling this issue):

|Player  |Playback|Behaviour
|--------|----------|-----------------------------------------------------------------------------
|**THEOplayer**|Works fine|**No retries**, gets newer segments<br/>*(You need to enable subtitles in the player)*
|**Shaka Player**|Works fine|**Retries once**, gets newer segments, and reports<br/>*(You need to enable subtitles in the player)*
|**DASH reference player**|Works fine|**Retries 3 times**, reports, stops subs<br/>*(Subtitles are enabled by default)*
|**Bitmovin**|Works fine|**Retries twice, gets next segment indefinitely**<br/>*(You need to enable subtitles in the player)*
|**Akamai**|Stutters, then works fine|**Keeps retrying same file indefinitely**<br/>*(Subtitles are enabled by default)*
|**JW Player**|Fails immediately|*(Subtitles are enabled by default)*
|**FlowPlayer**|Fails immediately|*(Subtitles are enabled by default)*

The list of test players is found here: https://github.com/add-ons/best-practices/wiki/Streaming-issues

**Comment by dagwieers** [comment]
> The players do MIND. They keep retrying the same file. They are not skipping retries on 415

I meant they don't mind the 415 (as in give up), **but they don't retry the same file**. THEOplayer handles it best as it continues supporting without affecting the playback or user and eventually picks up subtitles when they *are* available. IA both affects playback, as well as give up (on the subtitles). At least the playback eventually continues without any further issues, but without subtitles when they would be available.

> Something like this I like more than specifically looking for 415.

As I already mentioned, yes that was suggested. But again, you are looking at this from the perspective of this specific erratic behaviour of a stream that should not be, but I am looking from the perspective of streams we cannot influence and a player that should behave better despite errors. (resilience/reliability)

So even if this specific VOD provider would have fixed this, the player should handle such cases better.

**Comment by matthuisman** [comment]
And what is better to you?

How should IA handle errors?
You want retries but don't want the delay they introduce?

Only way I see to get that is threads.
But we need to wait for the initial video / audio (cant thread them)
But we don't want to wait for initial subs.

It seems to me - subs need to be treated differently.

With your file, does IA keep trying the subs during playback?

**Comment by dagwieers** [comment]
For the subtitle track retries are probably not very useful, but doing the next segment would be useful. So I guess the proposed PR by @mediaminister is exactly what we want *if* we cannot have threads and *if* it picks up newer segments.

So maybe the only thing missing from the PR is some guidance to why it was added, what would be better (threads?).

**Comment by matthuisman** [comment]
I think it's fine if we remove the check for 415 and instead check for any errror (not OK response)
as I said many comments back :)

I think != 200 may be wrong as you can have partial response codes etc.

Check if response code doesn't start with 2 should do it I think.
Redirects should already be handled by the time it gets here in code.

So, if it's a subtitle and response code not start with 2: don't retry.

UPDATE:
I've chatting with @glennguy  and he thinks there is a download thread in IA.
If it's just a single thread with a queue that won't help.
Unless instead of retrying, put the failed file back at the end of the queue to retry last.
Keeps getting put back into queue for X amount of times (retries)
That way, the first video and audio will get done, playback starts and subs keep retrying.

But if we can spawn a new thread for each download - then that should fix this issue.
Or, maybe there is something making IA wait for subs further on.
Or maybe it's Kodi that waits.

Did you confirm IA keeps trying the subs after playback starts? (which is what we want)
But maybe just not 10 times every time

**Comment by dagwieers** [comment]
I cannot confirm or deny because I have no environment to build Kodi at this time. Maybe @mediaminister can confirm in the morning.

**Comment by matthuisman** [comment]
don't need to build kodi. just have debug logging with curl enabled.
(actually, don't need curl enabled - IA will show the error)

Then playback the mpd.
You will be able to see the initial 10 retries.
Then hopefully see it keep retying every-time it re-downloads the mpd.

I'll check now :)

Also, what happens if you start the stream with subs disabled?
Does it start straight away?
If so - maybe it's more a kodi issue.
Its waiting for subs if they are enabled before starting playback.

Anyway, this issue has been a great way to start a discussion on retry logic in general and how things can be handled nicer. In conclusion - dont let subs affect playback.

**And by the way - it's pretty cool your provider supports and encourages Kodi and your add-on.**
Here in NZ / AU - it's usually no word until a dmca takedown (why I no longer public host my addon repos)

**Comment by dagwieers** [comment]
Ah, I figured you meant mediaminister's branch, because it affects the behaviour.

Yes, we have subtitles enabled by default, but this is by choice.

Again, we are not looking for a fix for our issue. We are looking into fixing Kodi's behaviour. In our add-on we switched to HLS for this specific stream which is not affected by the delay.

**Comment by dagwieers** [comment]
By coincidence I find this in my logs of 2 hours ago:
```
2020-05-11 23:08:20.078 T:1027597184   ERROR: AddOnLog: InputStream Adaptive: Decrypt Sample returns failure!
2020-05-11 23:09:56.450 T:985498496   ERROR: Previous line repeats 1 times.
2020-05-11 23:09:56.450 T:985498496   ERROR: AddOnLog: InputStream Adaptive: Download https://live-cf-vrt.akamaized.net/groupc/live/8edf3bdf-7db3-41c3-a318-72cb7f82de66/live.isml/dash/live-text_track_0_dut=1000-1589231389759.dash failed with error: 404
2020-05-11 23:09:57.888 T:1027597184   ERROR: AddOnLog: InputStream Adaptive: Decrypt Sample returns failure!
2020-05-11 23:10:34.771 T:985498496   ERROR: Previous line repeats 2 times.
2020-05-11 23:10:34.771 T:985498496   ERROR: AddOnLog: InputStream Adaptive: Download https://live-cf-vrt.akamaized.net/groupc/live/8edf3bdf-7db3-41c3-a318-72cb7f82de66/live.isml/dash/live-text_track_0_dut=1000-1589231428159.dash failed with error: 404
2020-05-11 23:10:51.362 T:1027597184   ERROR: AddOnLog: InputStream Adaptive: Decrypt Sample returns failure!
2020-05-11 23:12:12.709 T:985498496   ERROR: Previous line repeats 4 times.
2020-05-11 23:12:12.709 T:985498496   ERROR: AddOnLog: InputStream Adaptive: Download https://live-cf-vrt.akamaized.net/groupc/live/8edf3bdf-7db3-41c3-a318-72cb7f82de66/live.isml/dash/live-text_track_0_dut=1000-1589231526079.dash failed with error: 404
2020-05-11 23:13:06.994 T:1027597184   ERROR: AddOnLog: InputStream Adaptive: Decrypt Sample returns failure!
2020-05-11 23:14:34.706 T:985498496   ERROR: Previous line repeats 2 times.
2020-05-11 23:14:34.706 T:985498496   ERROR: AddOnLog: InputStream Adaptive: Download https://live-cf-vrt.akamaized.net/groupc/live/8edf3bdf-7db3-41c3-a318-72cb7f82de66/live.isml/dash/live-text_track_0_dut=1000-1589231668159.dash failed with error: 404
2020-05-11 23:14:45.801 T:1027597184   ERROR: AddOnLog: InputStream Adaptive: Decrypt Sample returns failure!
2020-05-11 23:15:18.895 T:985498496   ERROR: Previous line repeats 18 times.
```

Here we get 404 errors on subtitle tracks (this is a live tv channel that does have subtitle tracks) and here it doesn't seem to be retrying. Although it doesn't seem to affect playback either (possibly because it started fine with subtitles). Retries here are useless too.

This is the same VOD provider BTW...

**Comment by matthuisman** [comment]
ok, so 100% not just check for 415 then.

That is fine though - all the big players would done the same thing.
If it didn't affect playback - your happy right?
Only downside is log spam.
It needs to try the subs in case they do come back.
With a live edge - it may just be timing - so good to retry.

Good to see IA is at least trying the next text segment (not the same over and over)
Just needs to only do 1 try for subs though you reckon?
Think we need to stop saying retry - as thats ambiguous.
0 retries = 1 try then 0 retry?
So let's start saying "tries" as in how many total tries for that file.
For subs - we only want a single try?

And, you can download IA builds from PR's that are automatically built.
eg. https://jenkins.kodi.tv/blue/organizations/jenkins/peak3d%2Finputstream.adaptive/detail/PR-441/1/artifacts

**Comment by dagwieers** [comment]
Yes, I already added my thoughts to PR #441.

**Comment by dagwieers** [comment]
@matthuisman I updated the matrix of all player's behaviour here: https://github.com/peak3d/inputstream.adaptive/issues/433#issuecomment-626994942

And as you can see the detailed behaviour is also erratic for some players, despite not being affected in playback. Only THEOplayer handles this best IMO (if you consider 415 unrecoverable, which I think it is).

**Comment by matthuisman** [comment]
@dagwieers 
Try this: https://github.com/peak3d/inputstream.adaptive/pull/445

**Comment by dagwieers** [comment]
I found this while digging into HTTP 415: https://docs.unified-streaming.com/documentation/package/usage.html#fmp4-415

This gives alternative meanings to existing HTTP errors in a streaming context.

**Comment by matthuisman** [comment]
why are we back on 415...
You've already seen sometimes you get 404 errors as well.
And we have also confirmed a lot of players retry on it as well (don't specifically look for 415)

Its still the wrong error they are returning.
It's not invalid format or unsupported media - it's simply the file doesn't exist.

It means nothing - its just an obscure random error they have chosen to return for some reason.
If they return the correct 404 - behaviour everywhere will stay the same.

**Comment by dagwieers** [comment]
@matthuisman Because we did not understand why we got a **415 Unsupported Media Type**, and this confirms that HTTP 415 has different meanings, at least for Unified Streaming, but likely for others as well

*If you don't want to get this info, I suggest you disable notifications for this PR, because I am not writing this just for you, this is a log of our findings in general. Maybe someone else finds this useful.*

> It means nothing - its just an obscure random error they have chosen to return for some reason.

Point being, it is not an obscure random error. Thanks for reading that link.

**Comment by matthuisman** [comment]
the developer even told you thats a random error they chose.
So you know it's just them doing it.
It's not some sort of secret standard that no one knows about except this one crowd.

I read it - but they are using it incorrectly.
There file DOESNT EXIST. Its not invalid format, its not wrong media type.
Its a 404 - Does not exist!

415 would make sense if we requesting a mp4 stream and they only have a ts stream.

The fact you keep going away and trying to find different definitions for 415 is irrelevant as its the wrong error to return in the first place if the file doesnt exist.

And I am checking out.
You just seem to ignore any logic that doesn't fit your single stream.
Lets look for 415 errors because one stream uses it incorrectly.

I thought we already worked out to just check for any errors.
But back on 415 again. That is circles.

Those players you say work will just be checking for any error - not 415.
Open their source code and look for 415. It won't be there.
The error code is irrelevant.

**Comment by dagwieers** [comment]
@matthuisman We have never talked to a developer, and the guy who you think is a developer but is not, is not working on any products either, did not say it was a random error. In fact he said it was a very specific error: 

> Since the subtitle track is not (always) present, the origin can't provide the requested data, and returns the 415 status that you observed. From the point of view of our origin partner, this is expected behaviour.
>
>The player (TheoPlayer) we use on our official sites/apps ("touchpoints") handles this 415 as meaning "no subtitles present", and starts to play anyway. The players you tested (ffmpeg, jwplayer) behave differently. (Note that I'm trying to avoid pointing fingers at who's right or wrong)

I am not questioning anything we already established, please stop claiming I do.

**Comment by dagwieers** [comment]
And could you please stop arguing and maybe disable notifications? It is getting tiresome.

**Comment by matthuisman** [comment]
fine with me.
I look forward to seeing your fix

**Comment by mediaminister** [comment]
I looked into this again, 415 is returned because there are no subtitle samples created at the streaming backend of the content provider, the streaming provider [Unified Origin](https://docs.unified-streaming.com/documentation/vod/troubleshooting.html#unsupported-media-type) states the following in their documentation:

> In the error log of the webserver you will find 'FMP4_415 No samples for fragment', because the webserver tried to create a fragment and could not find samples to do so

I agree with @matthuisman that the content provider should fix this:

> If they have subs that come and go - then return files with no subs in them.
> This would allow subs to start half way through a segment as well.

Providing empty subtitle fragments is indeed the only right solution here.

> Have you asked the provider if they can just return a subs file with no subs in it?

No, but the content provider will try to implement this:

> but we're looking in to always adding a (possibly empty) subtitle track. If this can be done easily, we'll do that.

> Its still the wrong error they are returning.
> It's not invalid format or unsupported media - it's simply the file doesn't exist.

Indeed, this is an internal error between the content provider and the streaming provider. They shouldn't return 415 to the clients but instead a 404.

I'll close my pull request, adding more complexity to IA for a problem that is caused by the content provider is not a good idea.

**Comment by dagwieers** [comment]
Ok, I will open a PR to make AI more resilient. If Unified Origin would be returning a 404 instead of 415, IA has the exact same problem. And other players handle this more gracefully.

**Comment by peak3d** [comment]
Wow, this is a long discussion here :-) I suggest that IA callbacks the addon with the HTTP error code and your addon has the possibility to select what to do.

**Comment by dagwieers** [comment]
@peak3d I don't think that is necessary, as the add-on won't be able to implement what is ideal either. See #446 for the details. I am not sure if anyone wants a configurable behaviour for subtitle fragment errors.

**Comment by dagwieers** [comment]
@basilgello If your issue has not been fixed yet, could you open a new ticket for this?

**Comment by basilgello** [comment]
> @basilgello If your issue has not been fixed yet, could you open a new ticket for this?

Sure, will open today,
